ios/engine/KMEI/KeymanEngine/Classes/Resource Management/ResourceDownloadQueue.swift

```diff
@@ -9,82 +9,98 @@
+      case .simpleBatch(let batch):
+        return batch.resources
+      case .compositeBatch(let node):
+        return node.resources
+    }
   }
-  
-  public final var type: Resource
-  public final var resources: [AnyLanguageResource]?
+
+  var errors: [Error?] {
+    get {
+      switch(self) {
+        case .simpleBatch(let batch):
+          return batch.errors
+        case .compositeBatch(let node):
+          return node.errors
+      }
+    }
+
+    set(value) {
+      switch(self) {
+        case .simpleBatch(var batch):
+          batch.errors = value
+        case .compositeBatch(let node):
+          node.errors = value
+      }
+    }
+  }
+}
+
+protocol AnyDownloadTask {
+  var type: LanguageResourceType? { get }
+  var request: HTTPDownloadRequest { get }
+}
+
+class DownloadTask<Resource: LanguageResource>: AnyDownloadTask {
+  public final var type: LanguageResourceType?
+  public final var resources: [Resource]?
   public final var request: HTTPDownloadRequest
   
-  public init(do request: HTTPDownloadRequest, for resources: [AnyLanguageResource]?, type: DownloadTask.Resource) {
+  public init(do request: HTTPDownloadRequest, for resources: [Resource]?, type: LanguageResourceType?) {
     self.request = request
     self.type = type
     self.resources = resources
   }
 }
 
+enum DownloadActivityType {
+  case download, update
+}
+
+protocol AnyDownloadBatch {
+  var activity: DownloadActivityType { get }
+  var type: LanguageResourceType? { get }
+  var tasks: [AnyDownloadTask] { get }
+  var resources: [AnyLanguageResource] { get }
+  var errors: [Error?] { get set }
+}
+
 /**
  * Represents one overall resource-related command for requests against the Keyman Cloud API.
  */
-class DownloadBatch: DownloadNode {
-  public enum Activity {
-    case download, update, composite
-  }
-  
-  /*
-   * Three main cases so far:
-   * - [.download,  .keyboard]:      download new Keyboard (possibly with an associated lexical model)
-   * - [.download,  .lexicalModel]:  download new LexicalModel.
-   * - [.composite, .other]:   batch update of all resources, containing the following within its compositeQueue
-   *   - [.update, .keyboard]:      updates one Keyboard
-   *   - [.update, .lexicalModel]:  updates one LexicalModel
-   *   - Depending on needs, this may be extended to allow 'nested' .composite instances.
-   */
-  public final var activity: Activity
-  public final var type: DownloadTask.Resource
+class DownloadBatch<Resource: LanguageResource, Package: TypedKeymanPackage<Resource>>: AnyDownloadBatch {
+  public final var activity: DownloadActivityType
+  public final var type: LanguageResourceType?
   
-  public final var tasks: [DownloadNode]
+  public final var downloadTasks: [DownloadTask<Resource>]
   var errors: [Error?] // Only used by the ResourceDownloadQueue.
-  //public final var promise: Int
+  public final var completionBlock: ResourceDownloadManager.CompletionHandler<Resource, Package>? = nil
```

Comment:
Making this field definition possible was the primary reason for most of the rework.